File: exchange-processing-service/src/main/java/processing/service/ResolveUserService.java
package io.slurm.cources.processing.service;

import org.springframework.boot.autoconfigure.security.oauth2.resource.ResourceServerProperties;
import org.springframework.boot.autoconfigure.security.oauth2.resource.UserInfoRestTemplateFactory;
import org.springframework.cloud.sleuth.instrument.web.mvc.TracingClientHttpRequestInterceptor;
import org.springframework.security.oauth2.client.OAuth2RestTemplate;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@Service
public class ResolveUserService {
    private final UserInfoRestTemplateFactory restTemplateFactory;
    private final ResourceServerProperties resource;

    private final TracingClientHttpRequestInterceptor interceptor;

    public ResolveUserService(UserInfoRestTemplateFactory restTemplateFactory, ResourceServerProperties resource, TracingClientHttpRequestInterceptor interceptor) {
        this.restTemplateFactory = restTemplateFactory;
        this.resource = resource;
        this.interceptor = interceptor;
    }

    public Long resolveUserId() {
        OAuth2RestTemplate userInfoRestTemplate = restTemplateFactory.getUserInfoRestTemplate();

        userInfoRestTemplate.setInterceptors(List.of(interceptor));

        Map<String, Object> response = userInfoRestTemplate.getForEntity(resource.getUserInfoUri(), Map.class).getBody();
        return Optional.ofNullable(response)
                .map(r -> r.get("principal"))
                .map(Map.class::cast)
                .map(v -> (Integer) v.get("userId"))
                .map(Long::valueOf)
                .orElse(null);
    }
}

File: notification-bot/src/main/kotlin/notification/service/ResolveUserService.kt
package io.slurm.cources.notification.service

import org.springframework.boot.autoconfigure.security.oauth2.resource.ResourceServerProperties
import org.springframework.security.oauth2.client.DefaultOAuth2ClientContext
import org.springframework.security.oauth2.client.OAuth2ClientContext
import org.springframework.security.oauth2.client.OAuth2RestTemplate
import org.springframework.security.oauth2.client.token.grant.password.ResourceOwnerPasswordResourceDetails
import org.springframework.security.oauth2.common.DefaultOAuth2AccessToken
import org.springframework.stereotype.Service

@Service
class ResolveUserService(val details: ResourceOwnerPasswordResourceDetails, val resource: ResourceServerProperties) {
    fun resolveUserId(token: String): Long? {
        val accessToken = DefaultOAuth2AccessToken(token)
        val oauthCtx: OAuth2ClientContext = DefaultOAuth2ClientContext(accessToken)

        val restTemplate = OAuth2RestTemplate(details, oauthCtx)
        val response = restTemplate.getForEntity(resource.userInfoUri, Map::class.java).body

        return response?.let { r -> r["principal"] as MutableMap<*, *> }
            .let { v -> v?.get("userId") as Int? }
            .let { u -> u?.toLong() }
    }
}